client/packages/programs/src/JsonForms/components/ProgramEvent.tsx

```diff
@@ -0,0 +1,228 @@
+        }),
+      ])
+      .optional(),
+    documentType: z.string(),
+    eventType: z.string(),
+    /**
+     * Display option based on type.
+     */
+    display: z
+      .discriminatedUnion('type', [
+        z.object({ type: z.literal('number') }),
+        z.object({
+          type: z.literal('string'),
+          multiline: z.boolean().optional(),
+          rows: z.number().optional(),
+          show: z.array(
+            z
+              .tuple([z.string(), z.string().optional()])
+              .rest(z.string().optional())
+          ),
+        }),
+      ])
+      .optional(),
+  })
+  .strict();
+type Options = z.infer<typeof Options>;
+
+const extractAt = (datetime?: string, options?: Options): Date => {
+  if (!datetime) {
+    return new Date();
+  }
+  const beforeDate = new Date(new Date(datetime).getTime() - 1);
+  if (!options || !options.at) {
+    return beforeDate;
+  }
+
+  if (options.at.type === 'encounter') {
+    switch (options.at.encounterStartDatetime) {
+      case 'before':
+        return beforeDate;
+      case 'after':
+        return new Date(datetime);
+      default:
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        ((_: never) => {})(options.at.encounterStartDatetime as never);
+    }
+  } else if (options.at.type === 'programEnrolment') {
+    switch (options.at.programEnrolmentDatetime) {
+      case 'before':
+        return beforeDate;
+      case 'after':
+        return new Date(datetime);
+      default:
+        // eslint-disable-next-line @typescript-eslint/no-unused-vars
+        ((_: never) => {})(options.at.programEnrolmentDatetime as never);
+    }
+  }
+
+  return new Date();
+};
+
+const getDisplayOptions = (
+  data: string | null | undefined,
+  options?: Options
+) => {
+  const show =
+    options?.display?.type === 'string' ? options?.display?.show : null;
+
+  if (!show) {
+    return data;
+  }
+
+  const displayValue = show.find(value => value[0] === data)?.[1];
+  return displayValue ?? '';
+};
+
+const UIComponent = (props: ControlProps) => {
+  const { label, uischema, config } = props;
+  const [datetime, setDatetime] = React.useState<string | undefined>();
+  const patientId = config?.patientId;
+  const encounterDocName = config?.documentName.includes('Encounter');
+  const enrolmentDocName = config?.documentName.includes('Program');
+
+  const { data: currentEncounter } = useEncounter.document.byDocName(
+    encounterDocName ? config.documentName : undefined
+  );
+  const { data: programEnrolment } = useProgramEnrolments.document.byDocName(
+    enrolmentDocName ? config.documentName : undefined
+  );
+
+  const { errors, options } = useZodOptionsValidation(
+    Options,
+    uischema.options
+  );
+
+  useEffect(() => {
+    if (
+      options?.at?.type === 'encounter' &&
+      options?.at?.encounterStartDatetime
+    ) {
+      setDatetime(currentEncounter?.startDatetime);
```

Comment:
Nit: the fetching of the datetime field from the encounter/program could be closer to where the config is evaluated. For example, the config says which datetime should be used for the `at` value and it would be more natural to get the datetime in `extractAt`...

This would also work for the potential use-case where we want to calculate `at` based on a different field, e.g. `encounterEndDatetime`.

Reply:
[91cf53b](https://github.com/openmsupply/open-msupply/pull/1466/commits/91cf53bc3c4a95d3aaf182e1554500d23664a72d)